File: java/examples/erparticle/src/main/java/com/nedap/retail/services/examples/EpcisExample.java
package com.nedap.retail.services.examples;

import java.util.ArrayList;
import java.util.List;

import org.joda.time.DateTime;

import com.nedap.retail.messages.Client;
import com.nedap.retail.messages.epcis.v1_1.EpcisEvent;
import com.nedap.retail.messages.epcis.v1_1.EpcisEventContainer;
import com.nedap.retail.messages.epcis.v1_1.ObjectEvent;
import com.nedap.retail.messages.epcis.v1_1.cbv.Action;
import com.nedap.retail.messages.epcis.v1_1.cbv.Disposition;
import com.sun.jersey.api.client.UniformInterfaceException;

public class EpcisExample {

    public static void runExample(final Client client) {
        System.out.println("*** EPCIS API example ***");

        final EpcisEventContainer epcisEventsList = new EpcisEventContainer();
        epcisEventsList.events = createEvents();

        try {
            // capture
            System.out.println("Capturing epcis events...");
            client.captureEpcisEvents(epcisEventsList);
            System.out.println("--- Done ---");

        } catch (final UniformInterfaceException e) {
            System.err.println("Server responded with an error: " + e.getResponse().getEntity(String.class));
        }
    }

    private static List<EpcisEvent> createEvents() {
        final List<EpcisEvent> events = new ArrayList<>();
        events.add(createEpcisEvent1());
        events.add(createEpcisEvent2());
        return events;
    }

    private static EpcisEvent createEpcisEvent1() {
        final String id = "12f03260-c56f-11e3-9c1a-0800200c9a66";
        final DateTime eventTime = DateTime.parse("2014-02-01T12:00:00.000+01:00");
        final DateTime recordTime = DateTime.now();
        final String eventTimeZoneOffset = "+01:00";
        final Action action = Action.OBSERVE;
        final String bizLocation = "urn:epc:id:sgln:012345.67890.001";
        final String readPoint = "urn:epc:id:sgln:012345.67890.001";
        final Disposition disposition = Disposition.SELLABLE_ACCESSIBLE;

        final List<String> epcList = new ArrayList<>();
        epcList.add("urn:epc:id:sgtin:061414.12346.0001");
        epcList.add("urn:epc:id:sgtin:061414.12346.0002");
        epcList.add("urn:epc:id:sgtin:061414.12346.0003");
        epcList.add("urn:epc:id:sgtin:061414.12346.0004");

        final EpcisEvent event1 = new ObjectEvent(id, eventTime, recordTime, eventTimeZoneOffset, action, bizLocation,
                readPoint, disposition, epcList, null);

        return event1;
    }

    private static EpcisEvent createEpcisEvent2() {
        final String id = "12f03260-c56f-11e3-9c1a-0800200c9a67";
        final DateTime eventTime = DateTime.parse("2014-02-01T12:01:00.000+01:00");
        final DateTime recordTime = DateTime.now();
        final String eventTimeZoneOffset = "+01:00";
        final Action action = Action.OBSERVE;
        final String bizLocation = "urn:epc:id:sgln:012345.67890.002";
        final String readPoint = "urn:epc:id:sgln:012345.67890.001";
        final Disposition disposition = Disposition.SELLABLE_ACCESSIBLE;

        final List<String> epcList = new ArrayList<>();
        epcList.add("urn:epc:id:sgtin:061414.12346.0005");
        epcList.add("urn:epc:id:sgtin:061414.12346.0006");
        epcList.add("urn:epc:id:sgtin:061414.12346.0007");
        epcList.add("urn:epc:id:sgtin:061414.12346.0008");

        final EpcisEvent event2 = new ObjectEvent(id, eventTime, recordTime, eventTimeZoneOffset, action, bizLocation,
                readPoint, disposition, epcList, null);

        return event2;
    }
}

File: csharp/examples/NedapRetailApiExample/EpcisExample.cs
using Nedap.Retail.Api.Api;
using Nedap.Retail.Api.Client;
using Nedap.Retail.Api.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nedap.Retail.Api.Example
{
    internal class EpcisExample
    {
        public static void RunExample(Configuration configuration)
        {
            try
            {
                // capture
                EpcisApi epcisApi = new EpcisApi(configuration);
                Console.WriteLine("------------- Capture epcis events");
                List<EpcisEvent> events = CreateEvents();
                epcisApi.CaptureEpcisEvents(events);
                printEpcisEvents(events);
                Console.WriteLine("captured epcis events");

                // query
                Console.WriteLine("------------- Query epcis events");
                List<ParameterObject> parameters = new List<ParameterObject>();
                parameters.Add(new ParameterObject("EQ_read_point", null, null, new List<string>(new string[] { "urn:epc:id:sgln:012345.67890.001" })));
                EpcisQueryParameters queryParameters = new EpcisQueryParameters(parameters);

                List<EpcisEvent> queryResult = epcisApi.QueryEpcisEvents(queryParameters);
                Console.WriteLine("Got " + queryResult.Count + " epcis events:");
                foreach (EpcisEvent epcisEvent in queryResult)
                {
                    switch (epcisEvent.Type)
                    {
                        case EpcisEvent.TypeEnum.Objectevent:
                            Console.WriteLine("Object event with " + epcisEvent.EpcList.Count + " EPCs");
                            printEpcisEvent(epcisEvent);
                            break;
                        case EpcisEvent.TypeEnum.Aggregationevent:
                            Console.WriteLine("Aggregation event with " + epcisEvent.ChildEpcs.Count + " EPCs");
                            printEpcisEvent(epcisEvent);
                            break;
                        case EpcisEvent.TypeEnum.Transactionevent:
                            Console.WriteLine("Transaction event with " + epcisEvent.EpcList.Count + " EPCs");
                            printEpcisEvent(epcisEvent);
                            break;
                        case EpcisEvent.TypeEnum.Transformationevent:
                            Console.WriteLine("Transformation event with " + epcisEvent.InputEpcList.Count + " EPCs as input");
                            printEpcisEvent(epcisEvent);
                            break;
                    }
                }
            }
            catch (ApiException e)
            {
                Console.WriteLine(e.ToString());
            }

            Console.ReadKey();
        }

        private static List<EpcisEvent> CreateEvents()
        {
            List<EpcisEvent> events = new List<EpcisEvent>();

            events.Add(new EpcisEvent()
            {
                Type = EpcisEvent.TypeEnum.Objectevent,
                Id = new Guid("12f03260-c56f-11e3-9c1a-0800200c9a66").ToString(),                
                EventTime = DateTime.Now, //TODO: Time format
                EventTimeZoneOffset = "+01:00",
                RecordTime = DateTime.Now,
                EpcList = new List<string>()
                {
                    "urn:epc:id:sgtin:1234567.000246.0001",
                    "urn:epc:id:sgtin:1234567.000246.0002",
                    "urn:epc:id:sgtin:1234567.000246.0003",
                    "urn:epc:id:sgtin:1234567.000246.0004"
                },
                Action = EpcisEvent.ActionEnum.OBSERVE,
                BizStep = "urn:epcglobal:cbv:bizstep:cycle_counting",
                Disposition = "urn:epcglobal:cbv:disp:sellable_accessible",
                ReadPoint = "urn:epc:id:sgln:012345.67890.001",
                BizLocation = "urn:epc:id:sgln:012345.67890.001",
                BizTransactionList = new List<BizTransactionElement>() {
                    new BizTransactionElement() { BizTransaction = "123" }
                }
            });
            events.Add(new EpcisEvent()
            {
                Type = EpcisEvent.TypeEnum.Objectevent,
                Id = new Guid("12f03260-c56f-11e3-9c1a-0800200c9a67").ToString(),
                EventTime = DateTime.Now,
                EventTimeZoneOffset = "+01:00",
                RecordTime = DateTime.Now,
                EpcList = new List<string>()
                {
                    "urn:epc:id:sgtin:1234567.000246.0005",
                    "urn:epc:id:sgtin:1234567.000246.0006",
                    "urn:epc:id:sgtin:1234567.000246.0007",
                    "urn:epc:id:sgtin:1234567.000246.0008"
                },
                Action = EpcisEvent.ActionEnum.OBSERVE,
                BizStep = "urn:epcglobal:cbv:bizstep:cycle_counting",
                Disposition = "urn:epcglobal:cbv:disp:sellable_accessible",
                ReadPoint = "urn:epc:id:sgln:012345.67890.001",
                BizLocation = "urn:epc:id:sgln:012345.67890.002",
                BizTransactionList = new List<BizTransactionElement>() {
                    new BizTransactionElement() { BizTransaction = "456" }
                }
            });
            events.Add(new EpcisEvent()
            {
                Type = EpcisEvent.TypeEnum.Objectevent,
                Id = new Guid("12f03260-c56f-11e3-9c1a-0800200c9a68").ToString(),
                EventTime = DateTime.Now,
                EventTimeZoneOffset = "+01:00",
                RecordTime = DateTime.Now,
                EpcList = new List<string>()
                {
                    "urn:epc:id:sgtin:1234567.000246.0009",
                    "urn:epc:id:sgtin:1234567.000246.0010",
                    "urn:epc:id:sgtin:1234567.000246.0011",
                    "urn:epc:id:sgtin:1234567.000246.0012"
                },
                Action = EpcisEvent.ActionEnum.OBSERVE,
                BizStep = "urn:epcglobal:cbv:bizstep:cycle_counting",
                Disposition = "urn:epcglobal:cbv:disp:sellable_accessible",
                ReadPoint = "urn:epc:id:sgln:012345.67890.001",
                BizLocation = "urn:epc:id:sgln:012345.67890.001",
                BizTransactionList = new List<BizTransactionElement>() {
                    new BizTransactionElement() { BizTransaction = "123" }
                }
            });
            events.Add(new EpcisEvent()
            {
                Type = EpcisEvent.TypeEnum.Objectevent,
                Id = new Guid("12f03260-c56f-11e3-9c1a-0800200c9a69").ToString(),
                EventTime = DateTime.Now,
                EventTimeZoneOffset = "+01:00",
                RecordTime = DateTime.Now,
                EpcList = new List<string>()
                {
                    "urn:epc:id:sgtin:1234567.000246.0013",
                    "urn:epc:id:sgtin:1234567.000246.0014",
                    "urn:epc:id:sgtin:1234567.000246.0015",
                    "urn:epc:id:sgtin:1234567.000246.0016"
                },
                Action = EpcisEvent.ActionEnum.OBSERVE,
                BizStep = "urn:epcglobal:cbv:bizstep:cycle_counting",
                Disposition = "urn:epcglobal:cbv:disp:sellable_accessible",
                ReadPoint = "urn:epc:id:sgln:012345.67890.001",
                BizLocation = "urn:epc:id:sgln:012345.67890.002",
                BizTransactionList = new List<BizTransactionElement>() {
                    new BizTransactionElement() { BizTransaction = "456" }
                }
            });

            return events;
        }

        private static void printEpcisEvents(List<EpcisEvent> events)
        {
            Console.WriteLine("Retrieved EPCIS events within last minute:");
            for (int i = 0; i < events.Count; i++)
            {
                Console.WriteLine("\n");
                Console.WriteLine("\tEPCIS event " + (i + 1) + ";");
                printEpcisEvent(events[i]);
            }
        }

        private static void printEpcisEvent(EpcisEvent epcisEvent) {
            Console.WriteLine("\tId: " + epcisEvent.Id);
            Console.WriteLine("\tType: " + epcisEvent.Type);
            Console.WriteLine("\tEvent time: " + epcisEvent.EventTime);
            Console.WriteLine("\tAction: " + epcisEvent.Action);
            Console.WriteLine("\tDisposition: " + epcisEvent.Disposition);
            Console.WriteLine("\tBiz location: " + epcisEvent.BizLocation);
            Console.WriteLine("\tBiz step: " + epcisEvent.BizStep);
            Console.WriteLine("\tRead point: " + epcisEvent.ReadPoint);
            Console.WriteLine("\tEpc list: ");
            printEpcList(epcisEvent);
        }

        private static void printEpcList(EpcisEvent epcisEvent) {

            for (int i = 0; i < epcisEvent.EpcList.Count; i++) {
                Console.WriteLine("\t\t" + (i + 1) + " " + epcisEvent.EpcList[i]);
            }
        }   
    }
}